nixos/modules/programs/gnome-terminal.nix

```diff
@@ -0,0 +1,72 @@
+# GNOME Terminal.
+
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+
+  cfg = config.programs.gnome-terminal;
+
+  vteInitSnippet = ''
+    # Show current working directory in VTE terminals window title.
+    # Supports both bash and zsh, requires interactive shell.
+    . ${pkgs.gnome3.vte}/etc/profile.d/vte.sh
+  '';
+
+in
+
+{
+
+  # Added 2019-08-19
+  imports = [
+    (mkRenamedOptionModule
+      [ "services" "gnome3" "gnome-terminal-server" "enable" ]
+      [ "programs" "gnome-terminal" "enable" ])
+  ];
+
+  options = {
+
+    programs.gnome-terminal.enable = mkEnableOption "GNOME Terminal";
+
+    programs.bash.vteIntegration = mkOption {
+      default = false;
+      type = types.bool;
+      description = ''
+        Whether to enable Bash integration for VTE terminals.
+        This allows it to preserve the current directory of the shell
+        across terminals.
+      '';
+    };
+
+    programs.zsh.vteIntegration = mkOption {
+      default = false;
+      type = types.bool;
+      description = ''
+        Whether to enable Zsh integration for VTE terminals.
+        This allows it to preserve the current directory of the shell
+        across terminals.
+      '';
+    };
+
+  };
+
+  config = mkMerge [
+    (mkIf cfg.enable {
+      environment.systemPackages = [ pkgs.gnome3.gnome-terminal ];
+      services.dbus.packages = [ pkgs.gnome3.gnome-terminal ];
+      systemd.packages = [ pkgs.gnome3.gnome-terminal ];
+
+      # Bash is default
+      programs.bash.vteIntegration = true;
+    })
+
+    (mkIf config.programs.bash.vteIntegration {
+      programs.bash.interactiveShellInit = vteInitSnippet;
```

Comment:
How about just a single `programs.shell.vteIntegration` which then does
```nix
mkIf vteIntegration {
  programs.bash.interactiveShellInit = vteInitSnippet;
  programs.zsh.interactiveShellInit = vteInitSnippet;
}
```

The user already controls whether bash or zsh is enabled with `programs.{bash,zsh}.enable`

Reply:
> The user already controls whether bash or zsh is enabled with programs.{bash,zsh}.enable

Right, so even if `zsh` isn't enabled  `programs.zsh.interactiveShellInit` doesn't do anything unneeded.

However I feel like this would be slightly dissonant with the `programs.{bash,zsh}` interfaces, we could eliminate one extra option but I don't think `programs.shell.*` would have any more use.